src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1023,12 +1023,22 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
 
         public static Dictionary<TKey, TSource> ToDictionary<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector)
         {
-            return ToDictionary<TSource, TKey, TSource>(source, keySelector, IdentityFunction<TSource>.Instance, null);
+            return ToDictionary<TSource, TKey>(source, keySelector, null);
         }
 
         public static Dictionary<TKey, TSource> ToDictionary<TSource, TKey>(this IEnumerable<TSource> source, Func<TSource, TKey> keySelector, IEqualityComparer<TKey> comparer)
         {
-            return ToDictionary<TSource, TKey, TSource>(source, keySelector, IdentityFunction<TSource>.Instance, comparer);
+            if (source == null) throw Error.ArgumentNull("source");
+            if (keySelector == null) throw Error.ArgumentNull("keySelector");
+
+            int capacity = 0;
+
+            if (source is TSource[]) capacity = (source as TSource[]).Length; // Cast to Array is much faster than cast to ICollection
+            else if (source is ICollection<TSource>) capacity = (source as ICollection<TSource>).Count;
```

Comment:
> Cast to Array is extremely fast (if I remember correctly, it's just two comparisons). It is much much faster than to ICollection<T>.

The relative costs depend on the object you are trying to cast. Using quick comparison at https://gist.github.com/JonHanna/89a11548e216b0b93b98 the two compare thus as percentages:

Source|Relative cost of collection cast to array cast
----|----
`Enumerable.Repeat((object)null, 3)`|216.99
`Enumerable.Range(0, 2).Select(i => (object)i)`|182.11
`new object[0].Where(i => false)`|191.22
`new object[0]`|561.03
`new object[20]`|419.77
`new List<object>()`|81.28
`new HashSet<object>()`|96.80
`new List<string>()`|259.81
`new HashSet<string>()`|209.38
`new string[0]`|161.86
`new string[20]`|166.88
`new List<object>().AsReadOnly()`|111.21
`new List<string>().AsReadOnly()`|254.15

As seen casting to array can be even more expensive than `ICollection<T>`, and with the examples that implemented neither (likely the most common case by far) is more than half the cost.